rasa/core/policies/intent_ted_policy.py

```diff
@@ -536,6 +566,115 @@ def _check_unlikely_intent(
+        all trackers and categorize them into two buckets:
+            1. Similarity scores when `label_id` is the correct label.
+            2. Similarity scores when `label_id` is the wrong label.
+
+        Args:
+            output_scores: Model's predictions for each data point.
+            label_ids: Numerical IDs of labels for each data point.
+
+        Returns:
+            Both buckets of similarity scores grouped by each unique label id.
+        """
+        label_id_scores: Dict[int, List[List[float]]] = defaultdict(list)
+        unique_label_ids = np.unique(label_ids).tolist()
+        if LABEL_PAD_ID in unique_label_ids:
+            unique_label_ids.remove(LABEL_PAD_ID)
+
+        for label_id in unique_label_ids:
+            label_id_scores[label_id] = [[], []]
+
+        for index, all_pos_labels in enumerate(label_ids):
+
+            for candidate_label_id in unique_label_ids:
+                if candidate_label_id in all_pos_labels:
+                    label_id_scores[candidate_label_id][0].append(
+                        output_scores["similarities"][index, 0, candidate_label_id]
+                    )
+                else:
+                    label_id_scores[candidate_label_id][1].append(
+                        output_scores["similarities"][index, 0, candidate_label_id]
+                    )
+
+        # Get only unique scores so that duplicate
+        # trackers created because of permutations are pruned out.
+        unique_label_id_scores = {
+            label_id: (list(set(scores[0])), list(set(scores[1])))
+            for label_id, scores in label_id_scores.items()
+        }
+        return unique_label_id_scores
+
+    @staticmethod
+    def _compute_label_quantiles(
+        label_id_scores: Dict[int, Tuple[List[float], List[float]]]
+    ) -> Dict[int, List[float]]:
+        """Computes multiple quantiles for each label id.
+
+        The quantiles are computed over the negative scores
+        collected for each label id. However, no quantile score
+        can be greater than the minimum positive score collected
+        for the corresponding label id.
+
+        Args:
+            label_id_scores: Scores collected for each label id
+                over positive and negative trackers.
+
+        Returns:
+            Computed quantiles for each label id.
+        """
+        label_quantiles = {}
+
+        for label_id, (positive_scores, negative_scores) in label_id_scores.items():
+            label_quantiles[label_id] = [
+                min(
+                    min(positive_scores),
+                    np.quantile(
+                        negative_scores,
+                        1 - tolerance_value / 100.0,
+                        interpolation="lower",
+                    ),
+                )
+                if negative_scores
+                else min(positive_scores)
+                for tolerance_value in range(0, 100, 5)
```

Comment:
Perhaps the granularity of this calculation should be a variable/a policy parameter? We may want to have increments of 1 for instance.

Reply:
The if statement for a condition that never changes in the loop is a little hard to read. Why not:
```python
if negative_scores:
  # .. make quantile thresholds ..
else:
  # .. make default thresholds ..
```

Additionally, this for loop could be made into a vector op since np.quantile takes an array of quantiles to compute.

Reply:
> Perhaps the granularity of this calculation should be a variable/a policy parameter? 

I don't think we have enough evidence to show that people would like to customize this. In general, we try to reduce the number of configurable parameters as much as possible so that users don't need to think a lot about what parameters to tune. Moreover, adding new configuration parameters is always easier than removing one so I would prefer to take a more conservative approach to it until we have had more feedback from users. However, I think we should describe in the docs that `tolerance` can only be tuned in increments of `0.05` for now. Will take care of that in the docs PR. Wdyt?

Took care of the other comment on the restructuring the function. Thanks for the push 👍

Reply:
great points! 

> However, I think we should describe in the docs that tolerance can only be tuned in increments of 0.05 for now. 

I think that is more my concern, that we had this parameter advertised as continuous but under the hood it is discrete. I think documenting this primarily solves this issue.